runtime/metrics/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Runtime metric interface similar to native Prometheus metrics.
+//!
+//! This is intended to be used only for testing and debugging and **must never
+//! be used in production**. It requires the Substrate wasm tracing support
+//! and command line configuration: `--tracing-targets wasm_tracing=trace`.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+use primitives::v1::{RuntimeMetricLabelValues, RuntimeMetricLabels};
+
+#[cfg(not(feature = "std"))]
+use parity_scale_codec::Encode;
+
+#[cfg(not(feature = "std"))]
+use primitives::v1::{RuntimeMetricOp, RuntimeMetricRegisterParams, RuntimeMetricUpdate};
+
+#[cfg(not(feature = "std"))]
+pub struct CounterVec {
+	name: &'static str,
+	label_values: RuntimeMetricLabelValues,
+}
+
+#[cfg(not(feature = "std"))]
+impl CounterVec {
+	/// Create a new counter metric.
+	pub fn new(name: &'static str) -> Self {
+		Self::new_with_labels(name, "No description provided", RuntimeMetricLabels::default())
+	}
+
+	/// Create a new counter metric with specified `labels`.
+	pub fn new_with_labels(
+		name: &'static str,
+		description: &'static str,
+		labels: RuntimeMetricLabels,
+	) -> Self {
+		// Send a register metric operation to node side.
+		let metric_update = RuntimeMetricUpdate {
+			metric_name: sp_std::vec::Vec::from(name),
+			op: RuntimeMetricOp::Register(RuntimeMetricRegisterParams::new(
+				sp_std::vec::Vec::from(description),
+				labels,
+			)),
+		}
+		.encode();
+
+		// `from_utf8_unchecked` is safe, we only care about the `metric_name` which is static str.
```

Comment:
It is not obvious for me why this is safe. More specifically, why we can be sure that in a serialized representation of `RuntimeMetricUpdate` there will be no invalid UTF-8 sequences. At least we should update the safety proof.

However, I am wondering, why don't we use a safer version for this?

Reply:
Technically we could reduce this to be byte strings, and only try to parse on the node side into valid utf8 `String` types.

Reply:
We can be certain that a serialized representation of RuntimeMetricUpdate will not be a valid UTF-8 sequence. Actually, what we want si to `transmute` from u8 slice to str.

Reply:
I will update the comments.